File: dropbox-tray.js
/**
 * dropbox-tray - A simple Dropbox tray icon using NW.js
 * https://github.com/gavinhungry/dropbox-tray
 */

(function() {
  'use strict';

  var pkg = require('./package.json');
  var gui = require('nw.gui');

  var child_process = require('child_process');
  var path = require('path');
  var _ = require('underscore');
  _.str = require('underscore.string');

  /**
   * Resolve a dot-delimited string upon an object
   *
   * @param {Object} obj
   * @param {String} str
   * @return {Mixed}
   */
  var resolve = function(obj, str) {
    if (!str || !_.isString(str)) {
      return obj;
    }

    return _.reduce(str.split('.'), function(obj, prop) {
      return obj ? obj[prop] : null;
    }, obj);
  };

  var DROPBOX = {
    BIN: resolve(pkg, 'dropbox.bin'),

    ICON: {
      BUSY: resolve(pkg, 'dropbox.icons.busy') || 'img/busy.png',
      BUSY2: resolve(pkg, 'dropbox.icons.busy2') || 'img/busy2.png',
      ERROR: resolve(pkg, 'dropbox.icons.error') || 'img/error.png',
      IDLE: resolve(pkg, 'dropbox.icons.idle') || 'img/idle.png',
      OFFLINE: resolve(pkg, 'dropbox.icons.offline') ||  'img/offline.png'
    },

    STATUS: {
      IDLE: 'Up to date',
      INDEXING: 'Indexing',
      OFFLINE: 'Dropbox isn\'t running!',
      SYNCING: 'Syncing',
      UPLOADING: 'Uploading'
    },

    DIR: (function() {
      var infoPath = resolve(pkg, 'dropbox.info');
      if (!infoPath) {
        return;
      }

      var info = require(path.join(process.env.HOME, infoPath));
      return resolve(info, 'personal.path');
    })()
  };

  var dropbox = {
    tray: new gui.Tray({
      tooltip: 'Dropbox',
      icon: DROPBOX.ICON.OFFLINE
    }),

    /**
     * Execute a dropbox-cli command
     *
     * @param {String} option
     * @param {Array:String} [args]
     * @return {String}
     */
    cli: _.memoize(function(option, args) {
      var command = _.str.sprintf('%s %s %s', DROPBOX.BIN, option, (args || []).join(' '));

      try {
        return child_process.execSync(command).toString().trim();
      } catch(err) {
        return null;
      }
    }, function() {
      return _.flatten(arguments).join(' ');
    }),

    /**
     * Get an array of status messages
     *
     * @return {Array:String}
     */
    getStatuses: function() {
      return this.cli('status').split('\n');
    },

    /**
     * Get the primary status message
     *
     * @return {String}
     */
    getStatus: function() {
      var statuses = this.getStatuses();

      return _.find(statuses, function(status) {
        return _.str.startsWith(status, DROPBOX.STATUS.UPLOADING) ||
          _.str.startsWith(status, DROPBOX.STATUS.SYNCING) ||
          _.str.startsWith(status, DROPBOX.STATUS.INDEXING);
      }) || _.first(statuses);
    },

    /**
     * Check if Dropbox is running or not
     *
     * @return {Boolean} true if Dropbox is running, false otherwise
     */
    isRunning: function() {
      return this.getStatus() !== DROPBOX.STATUS.OFFLINE;
    },

    /**
     * Check if Dropbox is busy or idle
     *
     * @return {Boolean} true if Dropbox is busy, false if Dropbox is idle
     */
    isBusy: function() {
      return this.isRunning() && this.getStatus() !== DROPBOX.STATUS.IDLE;
    },

    /**
     * Flushes the cache of memoized functions
     */
    flush: function() {
      this.cli.cache = {};
    }
  };

  dropbox.tray.on('click', function() {
    gui.Shell.showItemInFolder(path.join(DROPBOX.DIR, '/.dropbox'));
  });

  setInterval(function() {
    var status = dropbox.getStatus();
    if (dropbox.tray.tooltip !== status) {
      dropbox.tray.tooltip = status;
    }

    if (dropbox.isBusy()) {
      if (dropbox.tray.icon === DROPBOX.ICON.BUSY) {
        dropbox.tray.icon = DROPBOX.ICON.BUSY2;
      } else {
        dropbox.tray.icon = DROPBOX.ICON.BUSY;
      }
    } else {
      dropbox.tray.icon = dropbox.isRunning() ? DROPBOX.ICON.IDLE : DROPBOX.ICON.OFFLINE;
    }
  }, 500);

  setInterval(function() {
    dropbox.flush();
  }, 2000);

})();

File: dropbox-tray.py
#!/usr/bin/env python3
#
# dropbox-tray: Dropbox tray icon
# https://github.com/gavinhungry/dropbox-tray
#

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, GObject

from enum import Enum
import json
import os
import signal
import subprocess

class DropboxTray:
  class Icon(Enum):
    OFFLINE = 'icons/offline.png'
    IDLE = 'icons/idle.png'
    BUSY = 'icons/busy.png'
    BUSY2 = 'icons/busy2.png'
    ERROR = 'icons/error.png'

  class Status(Enum):
    IDLE = 'Up to date'
    INDEXING = 'Indexing'
    OFFLINE = "Dropbox isn't running!"
    SYNCING = 'Syncing'
    UPLOADING = 'Uploading'

  def __init__(self):
    self.source = os.path.dirname(os.path.realpath(__file__))
    self.directory = self.getDirectory()
    self.initIcon()
    Gtk.main()

  def getDirectory(self):
    return json.load(open(os.path.expanduser('~/.dropbox/info.json')))['personal']['path']

  def openDirectory(self, widget):
    subprocess.Popen(['xdg-open', self.directory])

  def initIcon(self):
    self.icon = Gtk.StatusIcon()
    self.setIcon(self.Icon.OFFLINE)
    self.icon.connect('activate', self.openDirectory)
    self.icon.timeout = GObject.timeout_add(500, self.update)
    self.icon.set_visible(True)

  def setIcon(self, icon):
    self.icon.attr = icon
    self.icon.set_from_file(os.path.join(self.source, icon.value))

  def setTitle(self, title):
    if self.icon.get_tooltip_text() != title:
      self.icon.set_tooltip_text(title)

  def setIconBusy(self):
    self.setIcon(self.Icon.BUSY2 if self.icon.attr == self.Icon.BUSY else self.Icon.BUSY)

  def getStatuses(self):
    return list(filter(None, os.popen('dropbox-cli status').read().split('\n')))

  def getStatus(self):
    statuses = self.getStatuses()

    if not len (statuses):
      return 'Dropbox'

    filtered = list(filter(lambda status:
      str.startswith(status, self.Status.UPLOADING.value) or
      str.startswith(status, self.Status.SYNCING.value) or
      str.startswith(status, self.Status.INDEXING.value),
    statuses))

    return filtered[0] if len(filtered) else statuses[0]

  def isRunning(self):
    return self.getStatus() != self.Status.OFFLINE.value

  def isBusy(self):
    return self.isRunning() and self.getStatus() != self.Status.IDLE.value

  def update(self):
    self.setTitle(self.getStatus())

    if self.isBusy():
      self.setIconBusy()
    else:
      self.setIcon(self.Icon.IDLE if self.isRunning() else self.Icon.OFFLINE)

    return True

if __name__ == '__main__':
  signal.signal(signal.SIGINT, signal.SIG_DFL)
  DropboxTray()
